typed-racket-lib/typed-racket/logic/ineq.rkt

```diff
@@ -0,0 +1,920 @@
+                       (leq-negate (Leq->leq l2))))
+
+(define (Leq-implies-Leq? l1 l2)
+  (leq-imp-leq? (Leq->leq l1)
+                (Leq->leq l2)))
+
+
+;; satisfiable-Leqs?
+(define (satisfiable-Leqs? leqs)
+  (fme-sat? (Leqs->sli leqs)))
+
+;; Leqs-imply-Leq?
+(define (Leqs-imply-Leq-or-not-Leq? ls l)
+  (define leqs (Leqs->sli ls))
+  (define ineq (Leq->leq l))
+  (cond
+    [(fme-imp-leq? leqs ineq) #t]
+    [(fme-imp-leq? leqs (leq-negate ineq)) #f]
+    [else 'neither]))
+
+;; Leqs-imply-Leq?
+(define (Leqs-imply-Leq? ls l)
+  (fme-imp-leq? (Leqs->sli ls) (Leq->leq l)))
+
+;; Leqs-imply-not-Leq?
+(define (Leqs-imply-not-Leq? ls l)
+  (fme-imp-leq? (Leqs->sli ls) (leq-negate (Leq->leq l))))
+
+;; Leqs-imply-Leqs
+(define (Leqs-imply-Leqs? assumptions goals)
+  (fme-imp? (Leqs->sli assumptions)
+            (Leqs->sli goals)))
+
+
+;; *****************************************************
+;; helper functions
+
+
+;; Leqs->sli
+(define/cond-contract (Leqs->sli Leqs)
+  (-> (listof LeqProp?) sli?)
+  (map Leq->leq Leqs))
+
+;; Leq->leq
+(define (Leq->leq l)
+  (match l
+    [(LeqProp: (LExp: c1 ts1) (LExp: c2 ts2))
+     (leq (lexp c1 ts1) (lexp c2 ts2))]))
+
+
+;; *****************************************************************************
+;; Fourier-Motzkin elimination implementation
+;; http://en.wikipedia.org/wiki/Fourier-Motzkin_elimination
+;;
+;; Uses non-matrix representations of data
+;;
+;; Decides satisfiability, implication, etc... of/between systems of linear inequalities
+;; of the form ([(a_1x_1 + a_2x_2 + ... + a_c) <= (b_1x_1 + b_2x_2 + ... + b_c)] ...)
+;; *****************************************************************************
+
+
+(define (deduplicate xs)
+  (hash-keys (for/hash ([x (in-list xs)]) (values x #t))))
+
+(define-match-expander lexp:
+  (lambda (stx)
+    (syntax-case stx ()
+      [(_ const exp)
+       #'(cons const exp)])))
+
+(define-match-expander leq:
+  (lambda (stx)
+    (syntax-case stx ()
+      [(_ lhs rhs)
+       #'(cons lhs rhs)])))
+
+(define/cond-contract (lexp const terms)
+  (-> exact-integer?
+      (and/c (hash/c any/c (and/c exact-integer?
+                                  (not/c (=/c 0)))
+                     #:immutable #t)
+             hash-eq?)
+      lexp?)
+  (cons const terms))
+
+(define lexp-const car)
+(define lexp-terms cdr)
+
+(define-syntax-rule (lexp* t ...)
+  (list->lexp (list t ...)))
+
+(define/cond-contract (list->lexp terms)
+  (-> (listof (or/c exact-integer? (list/c exact-integer? any/c))) lexp?)
+  (let loop ([c 0]
+             [h (make-terms)]
+             [zxs terms])
+    (match zxs
+      [(list) (lexp c h)]
+      [(cons (list coeff var) zxs-rst)
+       (loop c (terms-set h var (+ coeff (terms-ref h var))) zxs-rst)]
+      [(cons constant zxs-rst)
+       (loop (+ constant c) h zxs-rst)])))
+
+; lexp-coeff
+(define/cond-contract (lexp-coeff l x)
+  (-> lexp? any/c exact-integer?)
+  (terms-ref (lexp-terms l) x))
+
+(define-syntax (check-true stx)
```

Comment:
I can put them in a test module. If they're in a test module, then it's fine if our test module depends on rackunit? (i.e. that won't add unwanted runtime deps, right?)